lib/src/model.dart

```diff
@@ -1961,38 +2187,37 @@ abstract class ModelElement implements Comparable, Nameable, Documentable {
   }
 
   String _calculateLinkedName() {
-    if (name.startsWith('_')) {
-      return HTML_ESCAPE.convert(name);
-    }
-    if (!(this is Method || this is Field) && !package.isDocumented(element)) {
+    // If we're calling this with an empty name, we probably have the wrong
+    // element associated with a ModelElement or there's an analysis bug.
+    assert(!name.isEmpty ||
+        (this.element is TypeDefiningElement &&
+            (this.element as TypeDefiningElement).type.name == "dynamic"));
+
+    if (isPrivate(element)) {
       return HTML_ESCAPE.convert(name);
     }
 
-    ModelElement c = (this is EnclosedElement)
-        ? (this as EnclosedElement).enclosingElement
-        : null;
-    if (c != null) {
-      if (!package.isDocumented(c.element)) {
-        return HTML_ESCAPE.convert(name);
-      }
-      if (c.name.startsWith('_')) {
-        return '${c.name}.${HTML_ESCAPE.convert(name)}';
-      }
+    if (href == null) {
+      package.warn(this, PackageWarning.noCanonicalFound);
+      return HTML_ESCAPE.convert(name);
     }
 
     var classContent = isDeprecated ? 'class="deprecated" ' : '';
     return '<a ${classContent}href="${href}">$name</a>';
   }
 
   // TODO(keertip): consolidate all the find library methods
+  // This differs from package.findOrCreateLibraryFor in a small way,
```

Comment:
I think the TODO is still relevant; the way these functions are reusing parts of each other's functionality suggests that the lines are wrong between them, even if it isn't obvious yet exactly how to draw them.

Renamed this method and tweaked comments in a couple related places to improve clarity.